crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -6772,13 +6775,10 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
                                 ConstraintKey::NarrowingConstraint(constraint),
                             ));
                         }
-                        EnclosingSnapshotResult::FoundBindings(bindings) => {
-                            if place_expr.is_symbol()
-                                && !enclosing_scope_id.is_function_like(db)
-                                && !is_immediately_enclosing_scope
-                            {
-                                continue;
-                            }
```

Comment:
Not so related to the main point of this PR, but I realized that this branch is actually unnecessary. For the condition here to be met, the enclosing scope needs to be non-function-like, i.e., class scope or module scope, and in that case, this case cannot be reached in the first place.
This is because in module scopes, snapshots are not recorded, and in class scopes, constraints are recorded instead of bindings.